### PR TITLE
Remove deprecated feature in 0.45 in unroller

### PR DESCRIPTION
**Pull request by henryzou50**
### Summary
Removes the class deprecated in 0.45 in qiskit/transpiler/passes/basis/unroller.py

### Details and comments
`Unroller` is going to be removed in Qiskit 1.0 and has been replaced by the `BasisTranslator` pass. However, using `BasisTranslator` as a direct replacement for `Unroller` is a bit unclear. `BasisTranslator` takes in an `EquivalenceLibrary` as one of its arguments, where `Unroller` does not. To adjust the tests, I mainly used:
`from qiskit.circuit.library.standard_gates.equivalence_library import (
    StandardEquivalenceLibrary as std_eqlib,
)`
as the `EquivalenceLibrary`. While most of the tests were successful, a few tests fail as there needs to be additional equivalence rules, so I am not sure what is the best way to approach this.

### CODE REVIEW

File: releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+upgrade:
+  - |
+    Removes the deprecated Unroller class in 'qiskit.transpiler.passes.basis'.
+    This class was deprecated in Qiskit Terra 0.45 and has been replaced by
+    `BasisTranslator`.
+
+    Note that `BasisTranslator` takes different arguments than Unroller, as it 
+    requires a `EquivalenceLibrary` object to be passed in.
```
Comment (Cryoris) [inline releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml]: It would be nice to add an example snippet of how to use the translator instead of the unroller, if you'd like to add one
```suggestion
    Removes the deprecated ``Unroller`` class in :mod:`qiskit.transpiler.passes.basis`.
    This class was deprecated in Qiskit 0.45 and has been replaced by
    :class:`.BasisTranslator`.

    Note that :class:`.BasisTranslator` takes different arguments than ``Unroller``, as it 
    requires an :class:`.EquivalenceLibrary` object to be passed in. 
```

File: qiskit/transpiler/preset_passmanagers/common.py [hunk 2]
```diff
@@ -447,9 +446,7 @@ def generate_translation_passmanager(
     Raises:
         TranspilerError: If the ``method`` kwarg is not a valid value
     """
-    if method == "unroller":
```
Comment (Cryoris) [inline qiskit/transpiler/preset_passmanagers/common.py:450]: With this removal, we also should remove the corresponding plugin, right? 

https://github.com/Qiskit/qiskit/blob/c799435fc1cc443e4701793930e68f7a07c7aa22/qiskit/transpiler/preset_passmanagers/builtin_plugins.py#L179

**Review by Cryoris — COMMENTED** [review]
Could you also remove the `"unroller"` translation method along with this PR? It's used e.g. in https://github.com/Qiskit/qiskit/blob/c799435fc1cc443e4701793930e68f7a07c7aa22/qiskit/compiler/transpiler.py#L344

**Comment by Cryoris** [comment]
> While most of the tests were successful, a few tests fail as there needs to be additional equivalence rules, so I am not sure what is the best way to approach this.

As you said, the basis translator works a bit differently than the unroller. So I'd say the new results don't have to match the previous ones 1:1, but if the result of the basis translator makes sense I think it's fine to update the test to the new output.

File: releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml (hunk 1 above)
Comment (henryzou50) [reply releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml]: Thanks for the feedback and sounds good I will make those changes.

File: qiskit/transpiler/preset_passmanagers/common.py (hunk 2 above)
Comment (henryzou50) [reply qiskit/transpiler/preset_passmanagers/common.py:450]: Yes that is correct, I will do that.

File: releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
+    :class:`.BasisTranslator`.
 
-    Note that `BasisTranslator` takes different arguments than Unroller, as it 
+    Note that :class:`.BasisTranslator` takes different arguments than ``Unroller``, as it 
     requires a `EquivalenceLibrary` object to be passed in. 
+
+    Example Usage:
+    Instead of using ``Unroller`` as follows:
+      Unroller(["u2"])
+    first create a EquivalenceLibrary like:
```
Comment (Cryoris) [inline releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml]:
```suggestion
    first create a :class:`.EquivalenceLibrary` like:
```
Also to format this as code you might have to write (note the double colon)
```python
first create an :class:`.EquivalenceLibrary` like::

    from qiskit. ... 
```

File: requirements.txt [hunk 4]
```diff
@@ -7,3 +7,6 @@ python-dateutil>=2.8.0
 stevedore>=3.0.0
 typing-extensions
 symengine>=0.11
+pillow>=8.0.0
+pydot>=1.4.1
+graphviz>=0.13.2
```
Comment (Cryoris) [inline requirements.txt]: It seems like these are leftovers from a merge, or are they on purpose? 🙂

File: test/python/circuit/test_controlled_gate.py [hunk 5]
```diff
@@ -923,11 +925,12 @@ def test_open_controlled_unitary_matrix(self, num_ctrl_qubits):
     def test_open_control_cx_unrolling(self):
         """test unrolling of open control gates when gate is in basis"""
         qc = QuantumCircuit(2)
-        qc.cx(0, 1, ctrl_state=0)
+        qc.append(U3Gate(np.pi, 0, np.pi), [0])
+        qc.cx(0, 1)
+        qc.append(U3Gate(np.pi, 0, np.pi), [0])
```
Comment (Cryoris) [inline test/python/circuit/test_controlled_gate.py]: This test doesn't seem to test what it was supposed to anymore: it should test that an open control CX gets decomposed into a closed control CX and X (or U) gates. What is the output it you apply the basis translator to the open control CX? The same question holds for the other open control tests below 🙂

File: test/python/circuit/test_controlled_gate.py [hunk 6]
```diff
@@ -1443,9 +1449,8 @@ def test_single_controlled_rotation_gates(self, gate, cgate):
         cqc = QuantumCircuit(self.num_ctrl + self.num_target)
         cqc.append(cgate, cqc.qregs[0])
         dag = circuit_to_dag(cqc)
-        with self.assertWarns(DeprecationWarning):
-            unroller = Unroller(["u", "cx"])
-        uqc = dag_to_circuit(unroller.run(dag))
+        basis_translator = BasisTranslator(std_eqlib, ["u", "cx", "ccu1", "ccrx", "ccry", "ccrz"])
```
Comment (Cryoris) [inline test/python/circuit/test_controlled_gate.py]: I think this might also not test the right thing anymore: by adding the operations such as `ccrz` to the basis translator, doesn't the basis translator just do nothing in the `cgrz` case? If the number of operations in the `u cx` basis increases using the basis translator instead of the unroller, we could change the test threshold (and might need to open an issue to investigate, if there's a signficant increase)

File: test/python/circuit/test_controlled_gate.py [hunk 7]
```diff
@@ -1466,9 +1471,8 @@ def test_composite(self):
         qc.append(self.grz.control(self.num_ctrl), qreg)
 
         dag = circuit_to_dag(qc)
-        with self.assertWarns(DeprecationWarning):
-            unroller = Unroller(["u", "cx"])
-        uqc = dag_to_circuit(unroller.run(dag))
+        basis_translator = BasisTranslator(std_eqlib, ["u", "cx", "ccrx", "ry", "ccry", "ccrz"])
```
Comment (Cryoris) [inline test/python/circuit/test_controlled_gate.py]: See comment above

File: test/python/qasm2/test_circuit_methods.py [hunk 8]
```diff
@@ -293,7 +298,7 @@ def test_from_qasm_str_custom_gate1(self):
         expected = QuantumCircuit(qr, name="circuit")
         expected.append(rinv, [qr[0]])
 
-        self.assertEqualUnroll(["sdg", "h"], circuit, expected)
+        self.assertEqualUnroll(["sdg", "h", "rinv"], circuit, expected)
```
Comment (Cryoris) [inline test/python/qasm2/test_circuit_methods.py]: I'm a bit confused about these changes. Wouldn't it be sufficient to change how `assertEqualUnroll` works? It seems here that the basis translator will simply not unroll the circuit anymore, would it?

File: requirements.txt (hunk 4 above)
Comment (henryzou50) [reply requirements.txt]: Yes they are leftovers and not on purpose, I will fix that. Thanks for the catch!

File: test/python/circuit/test_controlled_gate.py (hunk 5 above)
Comment (henryzou50) [reply test/python/circuit/test_controlled_gate.py]: I agree, the test does not seem to test what it was suppose to do anymore and I was a bit confused how to switch from Unroller to BasisTranslator in these cases. 

The output when applying the bassis translator to the open control CX is:

`qiskit.transpiler.exceptions.TranspilerError: "Unable to translate the operations in the circuit: ['cx_o0'] to the backend's (or manually specified) target basis: ['measure', 'delay', 'barrier', 'snapshot', 'cx', 'reset', 'u3']. This likely means the target basis is not universal or there are additional equivalence rules needed in the EquivalenceLibrary being used. For more details on this error see: https://docs.quantum-computing.ibm.com/api/qiskit/qiskit.transpiler.passes.BasisTranslator#translation-errors"`

File: test/python/circuit/test_controlled_gate.py (hunk 6 above)
Comment (henryzou50) [reply test/python/circuit/test_controlled_gate.py]: Hmmm, I was a bit confused here and that does make sense, but I am not currently sure how to change the test threshold. When I was attempting to convert this from Unroller to BasisTranslator, not including the operations such as `ccrz` would cause a similar `TranspilerError` like above. 

For instance, if we run `basis_translator = BasisTranslator(std_eqlib, ["u", "cx"])` instead, then the output for `test_single_controlled_rotation_gates_1` would be 
` qiskit.transpiler.exceptions.TranspilerError: "Unable to translate the operations in the circuit: ['ccu1'] to the backend's (or manually specified) target basis: ['snapshot', 'cx', 'delay', 'measure', 'barrier', 'u', 'reset']. This likely means the target basis is not universal or there are additional equivalence rules needed in the EquivalenceLibrary being used. For more details on this error see: https://docs.quantum-computing.ibm.com/api/qiskit/qiskit.transpiler.passes.BasisTranslator#translation-errors"`

File: test/python/qasm2/test_circuit_methods.py (hunk 8 above)
Comment (henryzou50) [reply test/python/qasm2/test_circuit_methods.py]: At first, I tried to only change how assertEqualUnroll works, but the transition from using `Unroller` to `BasisTranslator` was not as smooth as I thought. The basis translaator will not unroll many of the circuits with custom gates, and if we do not include the custom gate in the basis like `self.assertEqualUnroll(["sdg", "h"], circuit, expected)`, then the output will be 
`qiskit.transpiler.exceptions.TranspilerError: "Unable to translate the operations in the circuit: ['rinv'] to the backend's (or manually specified) target basis: ['snapshot', 'delay', 'h', 'measure', 'reset', 'sdg', 'barrier']. This likely means the target basis is not universal or there are additional equivalence rules needed in the EquivalenceLibrary being used. For more details on this error see: https://docs.quantum-computing.ibm.com/api/qiskit/qiskit.transpiler.passes.BasisTranslator#translation-errors"`

File: test/python/circuit/test_controlled_gate.py (hunk 5 above)
Comment (Cryoris) [reply test/python/circuit/test_controlled_gate.py]: Hmm yes the basis translator alone is not able to handle this right now, but then we can also not state that it is able to replace the unroller. Off the top of my head, we could either
* extend the basis translator to handle open controls (it should detect the `_o*` ending and apply corresponding `X` gates)
* use `transpile(circuit, basis_gates=..., optimization_level=0)` instead of the basis translator
* use the `generate_translation_passmanager` function to get a pass manager for translation

For this PR, the `transpile` is certainly the easiest approach, but let's double check with the rest of the team.

**Comment by henryzou50** [comment]
Adjusted the tests in `test_controlled_gate` and `test_circuit_methods` to use a combination of `BasisTranslator` and `UnrollCustomDefinitions` as previously some gates, such as open controls, were handled by `Unroller` but the `BasisTranslator` fails to handle them just by itself.

File: releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    
+    First create a EquivalenceLibrary like:
+      from qiskit.circuit.library.standard_gates.equivalence_library import (
+          StandardEquivalenceLibrary as std_eqlib,
+      )
+    and then use ``BasisTranslator`` and ``UnrollCustomDefinitions`` as follows:
+      pm = PassManager([
+          UnrollCustomDefinitions(std_eqlib, basis_gates)
+          BasisTranslator(std_eqlib, basis_gates),
+      ])
```
Comment (Cryoris) [inline releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml]: I'm not sure if that would currently render nicely, how about the following?
```suggestion
    Where previously ``Unroller(basis_gates)`` could be used,
    you can now use::
    
      from qiskit.circuit.library.standard_gates.equivalence_library import (
          StandardEquivalenceLibrary as std_eqlib,
      )
      pm = PassManager([
          UnrollCustomDefinitions(std_eqlib, basis_gates)
          BasisTranslator(std_eqlib, basis_gates),
      ])
      translated = pm.run(circuit)
```

File: requirements.txt [hunk 10]
```diff
@@ -6,4 +6,4 @@ dill>=0.3
 python-dateutil>=2.8.0
 stevedore>=3.0.0
 typing-extensions
-symengine>=0.11
```
Comment (Cryoris) [inline requirements.txt:9]: It might be better to revert this change so that the people looking for where `symengine>=0.11` was introduced don't come back to this PR 🙂

File: test/python/circuit/test_controlled_gate.py [hunk 11]
```diff
@@ -960,9 +967,9 @@ def test_open_control_ccx_unrolling(self):
         ccx = CCXGate(ctrl_state=0)
         qc.append(ccx, [0, 1, 2])
         dag = circuit_to_dag(qc)
-        with self.assertWarns(DeprecationWarning):
-            unroller = Unroller(["x", "ccx"])
-        unrolled_dag = unroller.run(dag)
+        unroller = UnrollCustomDefinitions(std_eqlib, ["x", "ccx"])
```
Comment (Cryoris) [inline test/python/circuit/test_controlled_gate.py]: How about introducing a seperate function for the equality check, like in the other test? I.e.
```
def assertEqualTranslated(self, circuit, unrolled_reference, basis):
        unroller = UnrollCustomDefinitions(std_eqlib, basis)
        basis_translator = BasisTranslator(std_eqlib, basis)
        unrolled = basis_translator(unroller(circuit))  
        self.assertEqual(unrolled, unrolled_reference)
```
Note that we can also skip the explicit conversion to `DAGCircuit`

**Review by Cryoris — COMMENTED** [review]
Thanks for the fixes, this now looks good. I left only two small comments below 🙂

File: test/python/circuit/test_controlled_gate.py (hunk 11 above)
Comment (henryzou50) [reply test/python/circuit/test_controlled_gate.py]: Sounds good, will do.

File: releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+---
+upgrade:
+  - |
+    Removes the deprecated ``Unroller`` class in :mod:`qiskit.transpiler.passes.basis`.
+    This class was deprecated in Qiskit Terra 0.45 and has been replaced by
```
Comment (mtreinish) [inline releasenotes/notes/remove-deprecated-unroller-4693330708c681e0.yaml:5]: Don't apply this suggestion as it's not worth wasting a CI round trip, we can fix this when we update the release notes for the final 1.0 release.

```suggestion
    This class was deprecated in Qiskit 0.45 and has been replaced by 
```

File: requirements.txt [hunk 13]
```diff
@@ -6,4 +6,4 @@ dill>=0.3
...
+symengine>=0.11
```
Comment (mtreinish) [inline requirements.txt:9]: For those staring at this for too long to figure out what the diff is (I pulled it locally and compared the files manually), this PR is removing a trailing line feed from this line. I guess github doesn't render this well.

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for removing this. I left 2 inline comments, one was a tiny fix in the release notes but we can fix that when we do the final release so there is no need to apply it. The other was just a comment for posterity about the diff github is highlighting in the requirements.txt file (there's no functional change to the file).
